arangod/IResearch/IResearchAqlAnalyzer.cpp

```diff
@@ -0,0 +1,516 @@
+
+  /// @brief whether or not a query is a modification query
+  virtual bool isModificationQuery() const noexcept override { return false; }
+
+  virtual bool isAsyncQuery() const noexcept override { return false; }
+
+  virtual void enterV8Context() override {
+    TRI_ASSERT(FALSE);
+    THROW_ARANGO_EXCEPTION_MESSAGE(
+        TRI_ERROR_NOT_IMPLEMENTED,
+        "CalculationQueryContext entering V8 context is not implemented");
+  }
+
+ private:
+  arangodb::aql::QueryOptions _queryOptions;
+  arangodb::CollectionNameResolver _resolver;
+  mutable CalculationTransactionContext _transactionContext;
+  std::unique_ptr<arangodb::transaction::Methods> _trx;
+};
+
+frozen::set<irs::string_ref, 4> forbiddenFunctions{"TOKENS", "NGRAM_MATCH",
+                                                   "PHRASE", "ANALYZER"};
+
+arangodb::Result validateQuery(std::string const& queryStringRaw, TRI_vocbase_t& vocbase) {
+  try {
+    CalculationQueryContext queryContext(vocbase);
+    auto queryString = arangodb::aql::QueryString(queryStringRaw);
+    auto ast = queryContext.ast();
+    TRI_ASSERT(ast);
+    Parser parser(queryContext, *ast, queryString);
+    parser.parse();
+    ast->validateAndOptimize(queryContext.trxForOptimization());
+    AstNode* astRoot = const_cast<AstNode*>(ast->root());
+    TRI_ASSERT(astRoot);
+    // Forbid all V8 related stuff as it is not available on DBServers where analyzers run.
+    if (ast->willUseV8()) {
+      return {TRI_ERROR_BAD_PARAMETER,
+              "V8 usage is forbidden for calculation analyzer"};
+    }
+
+    // no modification (as data access is forbidden) but to give more clear error message
+    if (ast->containsModificationNode()) {
+      return {TRI_ERROR_BAD_PARAMETER,
+              "DML is forbidden for calculation analyzer"};
+    }
+
+    // no traversal (also data access is forbidden) but to give mo clear error message
+    if (ast->containsTraversal()) {
+      return {TRI_ERROR_BAD_PARAMETER,
+              "Traversal usage is forbidden for calculation analyzer"};
+    }
+
+    std::string errorMessage;
+    // Forbid to use functions that reference analyzers -> problems on recovery as analyzers are not available for querying.
+    // Forbid all non-Dbserver runable functions as it is not available on DBServers where analyzers run.
+    arangodb::aql::Ast::traverseReadOnly(
+        ast->root(), [&errorMessage](arangodb::aql::AstNode const* node) -> bool {
+          switch (node->type) {
+            case arangodb::aql::NODE_TYPE_FCALL: {
+              auto func = static_cast<arangodb::aql::Function*>(node->getData());
+              if (!func->hasFlag(arangodb::aql::Function::Flags::CanRunOnDBServer) ||
+                  forbiddenFunctions.find(func->name) != forbiddenFunctions.end()) {
+                errorMessage = "Function '";
+                errorMessage.append(func->name)
+                    .append("' is forbidden for calculation analyzer");
```

Comment:
Several messages refer to a "calculation" analyzer, but in other places the analyzer is named "aql" analyzer.
I suggest we should be consistent with the naming.

Reply:
Good point. Initially working name was "calculation" but than it was decided to switch to "aql". But auto-replace seems to fail somewhere 🤦